src/Stratis.Bitcoin/Utilities/AsyncExecutionEvent.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Asynchronous event handler that can be registered with <see cref="AsyncExecutionEvent{TSender, TArg}/>.
+    /// </summary>
+    /// <typeparam name="TSender">Type of the sender object that is the source of the event.</typeparam>
+    /// <typeparam name="TArg">Type of the argument that is passed to the callback.</typeparam>
+    /// <param name="sender">Source of the event.</param>
+    /// <param name="arg">Callback argument.</param>
+    public delegate Task AsyncExecutionEventCallback<TSender, TArg>(TSender sender, TArg arg);
```

Comment:
[this can be declared as contravariant ](https://stackoverflow.com/questions/28687446/why-does-resharper-suggest-me-to-make-type-parameter-t-contravariant)
`<in TSender, in TArg>`

Reply:
not for now i guess, would need deeper look